### PR TITLE
Shared Flags for Build and BuildRun

### PR DESCRIPTION
**Pull request by otaviof**
Continue the work started on `flags` package to centralize all flag definitions in a single location, and using those on `shp build create`, `shp build run` and `shp buildrun create`, so they all rely in the same flags.

For instance:

```
$ shp build create --help

Creates a new Build instance using the informed first argument as its name. For example:

	$ shp build create my-app --source-url="..." --output-image="..."

Usage:
  shp build create <name> [flags]

Flags:
      --builder-credentials-secret string   name of the secret with builder-image pull credentials
      --builder-image string                image employed during the building process
      --dockerfile string                   path to dockerfile relative to repository
  -h, --help                                help for create
      --output-credentials-secret string    name of the secret with builder-image pull credentials
      --output-image string                 image employed during the building process
      --source-context-dir string           use a inner directory as context directory
      --source-credentials-secret string    name of the secret with git repository credentials
      --source-revision string              git repository source revision
      --source-url string                   git repository source URL
      --strategy-apiversion string          kubernetes api-version of the build-strategy resource (default "v1alpha1")
      --strategy-kind string                build-strategy kind (default "ClusterBuildStrategy")
      --strategy-name string                build-strategy name (default "buildpacks-v3")
      --timeout duration                    build process timeout
```

Additionally, improving doc-comments and getting started with long-descriptions.

### CODE REVIEW

**Comment by otaviof** [comment]
/retitle [WIP] Build and BuildRun Create Sub-commands

I'm going to have a look `shp build run` as well, make sure it's aligned.

**Comment by otaviof** [comment]
/retitle Shared Flags for Build and BuildRun

**Comment by alicerum** [comment]
/lgtm

this is awesome, thanks :D

File: pkg/shp/cmd/build/create.go [hunk 1]
```diff
@@ -1,106 +1,88 @@
...
-		},
-	}
-
-	createCommand.cmd.Flags().StringVarP(&createCommand.image, "output-image", "i", "", "Output image created by build")
-	createCommand.cmd.Flags().StringVarP(&createCommand.strategy, "strategy", "", "buildah", "Build strategy")
-	createCommand.cmd.Flags().StringVarP(&createCommand.image, "source-url", "u", "", "Source URL to run the build from")
-
-	createCommand.cmd.MarkFlagRequired("source-url")
+const buildCreateLongDesc = `
+Creates a new Build instance using the informed first argument as its name. For example:
```
Comment (adambkaplan) [inline pkg/shp/cmd/build/create.go:24]:
```suggestion
Creates a new Build instance using the first argument as its name. For example:
```

File: pkg/shp/cmd/build/create.go [hunk 2]
```diff
@@ -1,106 +1,88 @@
...
-	}
-
-	if c.image != "" {
-		c.build.Spec.Output = buildv1alpha1.Image{
-			Image: c.image,
-		}
+// Validate is used for user input validation of flags and other data.
+func (c *CreateCommand) Validate() error {
+	if c.name == "" {
+		return fmt.Errorf("name is not informed")
```
Comment (adambkaplan) [inline pkg/shp/cmd/build/create.go:48]:
```suggestion
		return fmt.Errorf("name must be provided")
```

File: pkg/shp/cmd/build/create.go [hunk 3]
```diff
@@ -1,106 +1,88 @@
...
+// Run executes the creation of a new Build instance using flags to fill up the details.
+func (c *CreateCommand) Run(params *params.Params, io *genericclioptions.IOStreams) error {
+	b := &buildv1alpha1.Build{Spec: *c.buildSpec}
+	flags.SanitizeBuildSpec(&b.Spec)
 
+	buildResource := resource.GetBuildResource(params)
+	if err := buildResource.Create(c.cmd.Context(), c.name, b); err != nil {
+		return err
+	}
+	fmt.Fprintf(io.Out, "Build created %q\n", c.name)
```
Comment (adambkaplan) [inline pkg/shp/cmd/build/create.go:62]:
```suggestion
	fmt.Fprintf(io.Out, "Created build %q\n", c.name)
```

File: pkg/shp/cmd/build/run.go [hunk 4]
```diff
@@ -5,72 +5,86 @@ import (
...
+	buildName    string                      // build name
+	buildRunSpec *buildv1alpha1.BuildRunSpec // stores command-line flags
 }
 
-// Cmd returns cobra command object
-func (c *RunCommand) Cmd() *cobra.Command {
-	return c.cmd
-}
+const buildRunLongDesc = `
+Creates a unique BuildRun instance for informed Build (first argument), which starts the build
```
Comment (adambkaplan) [inline pkg/shp/cmd/build/run.go:27]:
```suggestion
Creates a unique BuildRun instance for the given Build, which starts the build
```

File: pkg/shp/cmd/build/run.go [hunk 5]
```diff
@@ -5,72 +5,86 @@ import (
...
+process orchestrated by the Build-Controller. For example:
```
Comment (adambkaplan) [inline pkg/shp/cmd/build/run.go:28]:
```suggestion
process orchestrated by the Shipwright build controller. For example:
```

File: pkg/shp/cmd/build/run.go [hunk 6]
```diff
@@ -5,72 +5,86 @@ import (
...
 
-func runCmd() runner.SubCommand {
-	runCommand := &RunCommand{
-		cmd: &cobra.Command{
-			Use:   "run <name>",
-			Short: "Start a build specified by 'name'",
-			Args:  cobra.ExactArgs(1),
-		},
-	}
+	$ shp build create my-app --source-url="..." --output-image="..."
```
Comment (adambkaplan) [inline pkg/shp/cmd/build/run.go:30]: I think we can safely assume that the developer has created a Build already, therefore we don't need to reference the create command.

File: pkg/shp/cmd/buildrun/create.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+// CreateCommand reprents the build's create subcommand.
+type CreateCommand struct {
+	cmd *cobra.Command // cobra command instance
+
+	name         string                      // buildrun name
+	buildRunSpec *buildv1alpha1.BuildRunSpec // stores command-line flags
+}
+
+const buildRunCreateLongDesc = `
+Creates a new BuildRun instance using the informed argument as name, and requires --buildref-name to
```
Comment (adambkaplan) [inline pkg/shp/cmd/buildrun/create.go:24]:
```suggestion
Creates a new BuildRun instance using the given name, and requires --buildref-name to
```

File: pkg/shp/flags/build.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+import (
+	buildv1alpha1 "github.com/shipwright-io/build/pkg/apis/build/v1alpha1"
+	"github.com/spf13/pflag"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// BuildSpecFlags creates a BuildSpec instance based on command-line flags.
+func BuildSpecFlags(flags *pflag.FlagSet) *buildv1alpha1.BuildSpec {
```
Comment (adambkaplan) [inline pkg/shp/flags/build.go:11]: Style nit - make the function name a bit more descriptive:

```suggestion
// BuildSpecFromFlags creates a BuildSpec instance based on command-line flags.
func BuildSpecFromFlags(flags *pflag.FlagSet) *buildv1alpha1.BuildSpec {
```

File: pkg/shp/flags/build.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+	imageFlags(flags, "builder", spec.Builder)
+	imageFlags(flags, "output", &spec.Output)
+
+	timeoutFlags(flags, spec.Timeout)
+
+	return spec
+}
+
+// SanitizeBuildSpec checks for empty inner data structure and replaces them with nil.
```
Comment (adambkaplan) [inline pkg/shp/flags/build.go:48]: nit
```suggestion
// SanitizeBuildSpec checks for empty inner data structures and replaces them with nil.
```

File: pkg/shp/flags/buildrun.go [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+import (
+	buildv1alpha1 "github.com/shipwright-io/build/pkg/apis/build/v1alpha1"
+	"github.com/spf13/pflag"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// BuildRunSpecFlags BuildRun's spec represtantation as command-line flags.
+func BuildRunSpecFlags(flags *pflag.FlagSet) *buildv1alpha1.BuildRunSpec {
```
Comment (adambkaplan) [inline pkg/shp/flags/buildrun.go:12]: Style nit - make the function name and godoc more descriptive.

```suggestion
// BuildRunSpecFromFlags creates a BuildRun spec from command-line flags.
func BuildRunSpecFromFlags(flags *pflag.FlagSet) *buildv1alpha1.BuildRunSpec {
```

File: pkg/shp/flags/buildrun.go [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+	buildRefFlags(flags, spec.BuildRef)
+	serviceAccountFlags(flags, spec.ServiceAccount)
+	timeoutFlags(flags, spec.Timeout)
+	imageFlags(flags, "output", spec.Output)
+
+	return spec
+}
+
+// SanitizeBuildRunSpec when inner elements are empty, making sure they are replace by a nil pointer.
```
Comment (adambkaplan) [inline pkg/shp/flags/buildrun.go:30]: nit - clean up grammar in godoc and make it consistent with SanitizeBuildSpec.

```suggestion
// SanitizeBuildRunSpec checks for empty inner data structures and replaces them with nil.
```

**Review by adambkaplan — CHANGES_REQUESTED** [review]
Mainly grammar nits and doc clean up comments. I really like the refactoring in this PR, this will help make the CLI more maintainable.

File: pkg/shp/cmd/build/run.go (hunk 6 above)
Comment (otaviof) [reply pkg/shp/cmd/build/run.go]: Sure.

**Pull request merged**